docs/src/tutorials/implementing.md

```diff
@@ -20,6 +20,32 @@ MathOptInterface for a new solver.
     not answered by this guide, please ask them in the [Developer chatroom](https://gitter.im/JuliaOpt/JuMP-dev)
     so we can improve this guide!
 
+## A note on the API
+
+The API of MathOptInterface is large and varied. In order to support the
+diversity of solvers and use-cases, we make heavy use of
+[duck-typing](https://en.wikipedia.org/wiki/Duck_typing). That is, solvers are
+not expected to implement the full API, nor is there a well-defined minimal
+subset of what must be implemented. Instead, you should implement the API as
+necessary in order to make the solver function as you require.
+
+The main reason for using duck-typing over a well-defined API is that solvers
+work in different ways and target different use-cases.
+
+For example:
+
+ * Some solvers support incremental problem construction, support modification
+   after a solve, and have native support for things like variable names.
+ * Other solvers are "one-shot" solvers that require all of the problem data to
+   construct and solve the problem in a single function call. They do not
+   support modification or things like variable names.
+ * Other "solvers" are not solvers at all, but things like file readers. These
+   may only support functions like [`read_from_file`](@ref), and may not even
+   support the ability to add variables or constraints directly!
+ * Finally, some "solvers" are layers which take a problem as input, transform
+   it according to some rules, and pass the transformed problem to an inner
+   solver.
```

Comment:
How's this paragraph now

Reply:
It's accurate I think. New users feel the interface overwhelming when they need to implement it. This paragraph might improve things